File: src/main/java/com/aliiensmp/aliienResize/Config/Sizes.java
package com.aliiensmp.aliienResize.Config;

import com.aliiensmp.aliienResize.AliienResize;
import com.aliiensmp.aliienResize.Config.Records.CachedActionItem;
import com.aliiensmp.aliienResize.Config.Records.CachedSizeItem;
import com.aliiensmp.aliienResize.Config.Records.GuiData;
import com.aliiensmp.aliienResize.Config.Records.PriceData;
import com.aliiensmp.aliienResize.Config.Records.SizeNode;
import com.aliiensmp.aliienResize.Menus.MenuAction;
import com.aliiensmp.core.items.ItemBuilder;
import com.aliiensmp.core.lib.boostedyaml.YamlDocument;
import com.aliiensmp.core.lib.boostedyaml.block.implementation.Section;
import org.bukkit.Material;
import org.bukkit.inventory.ItemFlag;
import org.bukkit.inventory.ItemStack;

import java.util.*;

public class Sizes {

    // GUI settings cache
    public static String MENU_TITLE = "<dark_gray>Select your Size</dark_gray>";
    public static int MENU_ROWS = 6;
    public static Material MENU_LOCKED_MATERIAL = Material.BARRIER;
    public static List<String> MENU_DEFAULT_LORE = List.of();
    public static List<String> MENU_DEFAULT_LORE_NO_PERM = List.of();

    // Data caches
    public static final Map<String, SizeNode> SIZES_BY_ID = new LinkedHashMap<>();
    public static final Map<Integer, List<CachedSizeItem>> SIZE_ITEMS_BY_PAGE = new LinkedHashMap<>();
    public static final Map<Integer, List<CachedActionItem>> ACTION_ITEMS_BY_PAGE = new LinkedHashMap<>();

    public static int MENU_MAX_PAGE = 1;

    // Transport records for streams
    private record ParsedSizeEntry(SizeNode node, CachedSizeItem cachedItem, int page) {}
    private record RawActionItem(MenuAction action, Material material, String name, List<Integer> slots, List<String> lore, int modelData, List<ItemFlag> flags) {}

    /**
     * Rebuilds the runtime cache using functional streams, reading from both configs.
     */
    public static void loadFromConfigs(YamlDocument sizesConfig, YamlDocument mainMenuConfig, AliienResize plugin) {
        SIZES_BY_ID.clear();
        SIZE_ITEMS_BY_PAGE.clear();
        ACTION_ITEMS_BY_PAGE.clear();
        MENU_MAX_PAGE = 1;

        MENU_TITLE = mainMenuConfig.getString("menu-settings.title", "<dark_gray>Select your Size</dark_gray>");
        MENU_ROWS = mainMenuConfig.getInt("menu-settings.rows", 6);
        MENU_LOCKED_MATERIAL = parseMaterial(mainMenuConfig.getString("menu-settings.locked-material", "BARRIER"), Material.BARRIER, "locked-material", plugin);

        MENU_DEFAULT_LORE = Optional.ofNullable(sizesConfig.getStringList("defaults.lore")).orElse(List.of());
        MENU_DEFAULT_LORE_NO_PERM = Optional.ofNullable(sizesConfig.getStringList("defaults.lore-without-perm")).orElse(List.of());

        int maxSlots = MENU_ROWS * 9;

        // Parse size nodes
        Optional.ofNullable(sizesConfig.getSection("sizes")).ifPresentOrElse(
                section -> section.getRoutesAsStrings(false).stream()
                        .map(rawKey -> parseSizeEntry(section.getSection(rawKey), rawKey, maxSlots, plugin))
                        .filter(Optional::isPresent)
                        .map(Optional::get)
                        .forEach(Sizes::cacheSizeEntry),
                () -> plugin.getLogger().warning("No 'sizes' section found in sizes.yml.")
        );

        // Parse action items
        Optional.ofNullable(mainMenuConfig.getSection("items")).ifPresentOrElse(
                section -> section.getRoutesAsStrings(false).stream()
                        .map(rawKey -> parseRawActionItem(section.getSection(rawKey), rawKey, plugin))
                        .forEach(rawItem -> expandActionItem(rawItem, maxSlots)),
                () -> plugin.getLogger().info("No action items found in main-menu.yml.")
        );

        // Sort arrays for predictable rendering
        SIZE_ITEMS_BY_PAGE.values().forEach(items -> items.sort(Comparator.comparingInt(CachedSizeItem::slot)));
        ACTION_ITEMS_BY_PAGE.values().forEach(items -> items.sort(Comparator.comparingInt(CachedActionItem::slot)));

        plugin.getLogger().info("Loaded " + SIZES_BY_ID.size() + " sizes across " + MENU_MAX_PAGE + " page(s).");
    }

    private static Optional<ParsedSizeEntry> parseSizeEntry(Section section, String rawKey, int maxSlots, AliienResize plugin) {
        String id = rawKey.toLowerCase(Locale.ROOT);

        int slot = section.getInt("gui.slot", 0);
        if (slot < 0 || slot >= maxSlots) {
            plugin.getLogger().warning("Skipping size '" + id + "': slot " + slot + " is outside GUI bounds.");
            return Optional.empty();
        }

        double scale = section.getDouble("scale", 1.0);
        String permission = section.getString("permission", "");
        int page = Math.max(1, section.getInt("gui.page", 1));

        List<String> specificLore = Optional.ofNullable(section.getStringList("gui.lore")).orElse(List.of());
        List<String> specificNoPermLore = Optional.ofNullable(section.getStringList("gui.lore-without-perm")).orElse(List.of());

        // Fallback to defaults if specific lore is empty
        List<String> rawLore = specificLore.isEmpty() ? MENU_DEFAULT_LORE : specificLore;
        List<String> rawNoPermLore = specificNoPermLore.isEmpty() ? MENU_DEFAULT_LORE_NO_PERM : specificNoPermLore;

        List<ItemFlag> itemFlags = parseItemFlags(section.getStringList("gui.item-flags"), id);
        Material material = parseMaterial(section.getString("gui.material", "STONE"), Material.STONE, id, plugin);
        String name = section.getString("gui.name", "<green>" + rawKey);
        int modelData = section.getInt("gui.model-data", 0);

        boolean purchasable = section.getBoolean("price.enabled", false);
        String currency = section.getString("price.currency", "VAULT");
        double amount = section.getDouble("price.amount", 0.0);

        // Format placeholders
        String priceText = (Math.rint(amount) == amount) ? String.valueOf((long) amount) : String.valueOf(amount);
        String scaleText = String.valueOf(scale);
        String suffixText = purchasable ? plugin.getCurrencyManager().getSuffix(id) : "";

        List<String> availableLore = replacePlaceholders(rawLore, "%price%", priceText, "%scale%", scaleText, "%permission%", permission, "%suffix%", suffixText);
        List<String> noPermLore = replacePlaceholders(rawNoPermLore, "%price%", priceText, "%scale%", scaleText, "%permission%", permission, "%suffix%", suffixText);

        ItemFlag[] flagsArray = itemFlags.toArray(ItemFlag[]::new);

        ItemStack availableItem = buildItem(material, name, availableLore, modelData, flagsArray, false);
        ItemStack selectedItem = buildItem(material, name, availableLore, modelData, flagsArray, true);
        ItemStack noPermItem = buildItem(MENU_LOCKED_MATERIAL, name, noPermLore, modelData, flagsArray, false);

        GuiData guiData = new GuiData(material.name(), slot, page, name, specificLore, specificNoPermLore, modelData, itemFlags);
        PriceData priceData = new PriceData(purchasable, currency, amount);
        SizeNode sizeNode = new SizeNode(id, scale, permission, guiData, priceData);
        CachedSizeItem cachedItem = new CachedSizeItem(slot, id, permission, scale, availableItem, selectedItem, noPermItem);

        return Optional.of(new ParsedSizeEntry(sizeNode, cachedItem, page));
    }

    private static RawActionItem parseRawActionItem(Section section, String rawKey, AliienResize plugin) {
        MenuAction parsedAction;
        try {
            parsedAction = MenuAction.valueOf(section.getString("action", "NONE").toUpperCase(Locale.ROOT));
        } catch (IllegalArgumentException e) {
            plugin.getLogger().warning("Invalid action in main-menu.yml for item '" + rawKey + "'. Defaulting to NONE.");
            parsedAction = MenuAction.NONE;
        }

        return new RawActionItem(
                parsedAction,
                parseMaterial(section.getString("material", "STONE"), Material.STONE, rawKey, plugin),
                section.getString("name", " "),
                section.getIntList("slots"),
                Optional.ofNullable(section.getStringList("lore")).orElse(List.of()),
                section.getInt("model-data", 0),
                parseItemFlags(section.getStringList("item-flags"), rawKey)
        );
    }

    private static void expandActionItem(RawActionItem rawItem, int maxSlots) {
        ItemFlag[] flagsArray = rawItem.flags().toArray(new ItemFlag[0]);

        for (int page = 1; page <= MENU_MAX_PAGE; page++) {
            if (!shouldDisplayOnPage(rawItem.action(), page)) continue;

            int targetPage = resolveTargetPage(rawItem.action(), page);
            String targetPageText = String.valueOf(targetPage);

            String resolvedName = rawItem.name().replace("%target_page%", targetPageText);
            List<String> resolvedLore = replacePlaceholders(rawItem.lore(), "%target_page%", targetPageText, "", "", "", "", "", "");
            ItemStack cachedItem = buildItem(rawItem.material(), resolvedName, resolvedLore, rawItem.modelData(), flagsArray, false);

            for (int slot : rawItem.slots()) {
                if (slot >= 0 && slot < maxSlots) {
                    ACTION_ITEMS_BY_PAGE.computeIfAbsent(page, k -> new ArrayList<>()).add(
                            new CachedActionItem(slot, rawItem.action(), targetPage, cachedItem)
                    );
                }
            }
        }
    }

    private static void cacheSizeEntry(ParsedSizeEntry entry) {
        SIZES_BY_ID.put(entry.node().id(), entry.node());
        SIZE_ITEMS_BY_PAGE.computeIfAbsent(entry.page(), k -> new ArrayList<>()).add(entry.cachedItem());
        MENU_MAX_PAGE = Math.max(MENU_MAX_PAGE, entry.page());
    }

    private static boolean shouldDisplayOnPage(MenuAction action, int page) {
        return switch (action) {
            case NEXT_PAGE -> page < MENU_MAX_PAGE;
            case PREVIOUS_PAGE -> page > 1;
            default -> true;
        };
    }

    private static int resolveTargetPage(MenuAction action, int page) {
        return switch (action) {
            case NEXT_PAGE -> page + 1;
            case PREVIOUS_PAGE -> page - 1;
            default -> page;
        };
    }

    private static Material parseMaterial(String name, Material fallback, String id, AliienResize plugin) {
        return Optional.ofNullable(name)
                .filter(n -> !n.isBlank())
                .map(n -> Material.matchMaterial(n.toUpperCase(Locale.ROOT)))
                .orElseGet(() -> {
                    plugin.getLogger().warning("Invalid material '" + name + "' for '" + id + "'. Using " + fallback.name() + ".");
                    return fallback;
                });
    }

    private static List<ItemFlag> parseItemFlags(List<String> flags, String id) {
        return Optional.ofNullable(flags).orElse(List.of()).stream()
                .map(flag -> {
                    try {
                        return ItemFlag.valueOf(flag.toUpperCase(Locale.ROOT));
                    } catch (IllegalArgumentException e) {
                        return null;
                    }
                })
                .filter(Objects::nonNull)
                .toList();
    }

    private static List<String> replacePlaceholders(List<String> list, String p1, String v1, String p2, String v2, String p3, String v3, String p4, String v4) {
        return list.stream()
                .map(line -> line.replace(p1, v1).replace(p2, v2).replace(p3, v3).replace(p4, v4))
                .toList();
    }

    private static ItemStack buildItem(Material material, String name, List<String> lore, int modelData, ItemFlag[] flags, boolean glow) {
        ItemBuilder builder = new ItemBuilder(material)
                .name(name)
                .stringLore(lore)
                .customModelData(modelData)
                .addFlags(flags)
                .glow(glow);

        return builder.build();
    }
}

File: src/main/kotlin/com/aliiensmp/aliienResize/config/Sizes.kt
package com.aliiensmp.aliienResize.config

import com.aliiensmp.aliienResize.AliienResize
import com.aliiensmp.aliienResize.config.data.*
import com.aliiensmp.aliienResize.Menus.MenuAction
import com.aliiensmp.core.config.Key
import com.aliiensmp.core.items.ItemBuilder
import com.aliiensmp.core.lib.boostedyaml.YamlDocument
import com.aliiensmp.core.lib.boostedyaml.block.implementation.Section
import org.bukkit.Material
import org.bukkit.inventory.ItemFlag
import org.bukkit.inventory.ItemStack
import java.util.Locale
import kotlin.math.round

object Sizes {

    @Key("menu-settings.title")
    var MENU_TITLE: String = "<dark_gray>Select your Size</dark_gray>"

    @Key("menu-settings.rows")
    var MENU_ROWS: Int = 6

    @Key("menu-settings.locked-material")
    private var RAW_LOCKED_MATERIAL: String = "BARRIER"

    val MENU_LOCKED_MATERIAL: Material
        get() = Material.matchMaterial(RAW_LOCKED_MATERIAL.uppercase(Locale.ROOT)) ?: Material.BARRIER

    @Key("defaults.lore")
    var MENU_DEFAULT_LORE: List<String> = emptyList()

    @Key("defaults.lore-without-perm")
    var MENU_DEFAULT_LORE_NO_PERM: List<String> = emptyList()

    val SIZES_BY_ID = linkedMapOf<String, SizeNode>()
    val SIZE_ITEMS_BY_PAGE = linkedMapOf<Int, MutableList<CachedSizeItem>>()
    val ACTION_ITEMS_BY_PAGE = linkedMapOf<Int, MutableList<CachedActionItem>>()

    var MENU_MAX_PAGE: Int = 1
        private set

    private data class ParsedSizeEntry(val node: SizeNode, val cachedItem: CachedSizeItem, val page: Int)
    private data class RawActionItem(
        val action: MenuAction,
        val material: Material,
        val name: String,
        val slots: List<Int>,
        val lore: List<String>,
        val modelData: Int,
        val flags: List<ItemFlag>
    )

    /**
     * Rebuilds the runtime cache
     */
    fun loadFromConfigs(sizesConfig: YamlDocument, mainMenuConfig: YamlDocument, plugin: AliienResize) {
        SIZES_BY_ID.clear()
        SIZE_ITEMS_BY_PAGE.clear()
        ACTION_ITEMS_BY_PAGE.clear()
        MENU_MAX_PAGE = 1

        val maxSlots = MENU_ROWS * 9

        // Parse size nodes
        sizesConfig.getSection("sizes")?.let { section ->
            section.getRoutesAsStrings(false)
                .mapNotNull { rawKey -> parseSizeEntry(section.getSection(rawKey), rawKey, maxSlots, plugin) }
                .forEach { cacheSizeEntry(it) }
        } ?: plugin.logger.warning("No 'sizes' section found in sizes.yml.")

        // Parse action items
        mainMenuConfig.getSection("items")?.let { section ->
            section.getRoutesAsStrings(false)
                .map { rawKey -> parseRawActionItem(section.getSection(rawKey), rawKey, plugin) }
                .forEach { expandActionItem(it, maxSlots) }
        } ?: plugin.logger.info("No action items found in main-menu.yml.")

        // Sort lists
        SIZE_ITEMS_BY_PAGE.values.forEach { items -> items.sortBy { it.slot } }
        ACTION_ITEMS_BY_PAGE.values.forEach { items -> items.sortBy { it.slot } }

        plugin.logger.info("Loaded ${SIZES_BY_ID.size} sizes across $MENU_MAX_PAGE page(s).")
    }

    private fun parseSizeEntry(section: Section, rawKey: String, maxSlots: Int, plugin: AliienResize): ParsedSizeEntry? {
        val id = rawKey.lowercase(Locale.ROOT)
        val slot = section.getInt("gui.slot", 0)

        if (slot !in 0..<maxSlots) {
            plugin.logger.warning("Skipping size '$id': slot $slot is outside GUI bounds.")
            return null
        }

        val scale = section.getDouble("scale", 1.0)
        val permission = section.getString("permission", "")
        val page = section.getInt("gui.page", 1).coerceAtLeast(1)

        val specificLore = section.getStringList("gui.lore") ?: emptyList()
        val specificNoPermLore = section.getStringList("gui.lore-without-perm") ?: emptyList()

        // Fallbacks
        val rawLore = specificLore.ifEmpty { MENU_DEFAULT_LORE }
        val rawNoPermLore = specificNoPermLore.ifEmpty { MENU_DEFAULT_LORE_NO_PERM }

        val itemFlags = parseItemFlags(section.getStringList("gui.item-flags"))
        val material = parseMaterial(section.getString("gui.material", "STONE"), Material.STONE, id, plugin)
        val name = section.getString("gui.name", "<green>$rawKey")
        val modelData = section.getInt("gui.model-data", 0)

        val purchasable = section.getBoolean("price.enabled", false)
        val currency = section.getString("price.currency", "VAULT")
        val amount = section.getDouble("price.amount", 0.0)

        // String evaluations
        val priceText = if (round(amount) == amount) amount.toLong().toString() else amount.toString()
        val scaleText = scale.toString()
        val suffixText = if (purchasable) plugin.currencyManager.getSuffix(id) else ""

        val availableLore = replacePlaceholders(rawLore, priceText, scaleText, permission, suffixText)
        val noPermLore = replacePlaceholders(rawNoPermLore, priceText, scaleText, permission, suffixText)

        val flagsArray = itemFlags.toTypedArray()
        val availableItem = buildItem(material, name, availableLore, modelData, flagsArray, false)
        val selectedItem = buildItem(material, name, availableLore, modelData, flagsArray, true)
        val noPermItem = buildItem(MENU_LOCKED_MATERIAL, name, noPermLore, modelData, flagsArray, false)

        val guiData = GuiData(material.name, slot, page, name, specificLore, specificNoPermLore, modelData, itemFlags)
        val priceData = PriceData(purchasable, currency, amount)
        val sizeNode = SizeNode(id, scale, permission, guiData, priceData)
        val cachedItem = CachedSizeItem(slot, id, permission, scale, availableItem, selectedItem, noPermItem)

        return ParsedSizeEntry(sizeNode, cachedItem, page)
    }

    private fun parseRawActionItem(section: Section, rawKey: String, plugin: AliienResize): RawActionItem {
        val actionStr = section.getString("action", "NONE").uppercase(Locale.ROOT)
        val parsedAction = runCatching { MenuAction.valueOf(actionStr) }.getOrElse {
            plugin.logger.warning("Invalid action in main-menu.yml for item '$rawKey'. Defaulting to NONE.")
            MenuAction.NONE
        }

        return RawActionItem(
            action = parsedAction,
            material = parseMaterial(section.getString("material", "STONE"), Material.STONE, rawKey, plugin),
            name = section.getString("name", " "),
            slots = section.getIntList("slots") ?: emptyList(),
            lore = section.getStringList("lore") ?: emptyList(),
            modelData = section.getInt("model-data", 0),
            flags = parseItemFlags(section.getStringList("item-flags"))
        )
    }

    private fun expandActionItem(rawItem: RawActionItem, maxSlots: Int) {
        val flagsArray = rawItem.flags.toTypedArray()

        for (page in 1..MENU_MAX_PAGE) {
            if (!shouldDisplayOnPage(rawItem.action, page)) continue

            val targetPage = resolveTargetPage(rawItem.action, page)
            val targetPageText = targetPage.toString()

            val resolvedName = rawItem.name.replace("%target_page%", targetPageText)
            val resolvedLore = replacePlaceholders(rawItem.lore, "", "", "", "", targetPageText)
            val cachedItem = buildItem(rawItem.material, resolvedName, resolvedLore, rawItem.modelData, flagsArray, false)

            for (slot in rawItem.slots) {
                if (slot in 0..<maxSlots) {
                    ACTION_ITEMS_BY_PAGE.computeIfAbsent(page) { mutableListOf() }.add(
                        CachedActionItem(slot, rawItem.action, targetPage, cachedItem)
                    )
                }
            }
        }
    }

    private fun cacheSizeEntry(entry: ParsedSizeEntry) {
        SIZES_BY_ID[entry.node.id] = entry.node
        SIZE_ITEMS_BY_PAGE.computeIfAbsent(entry.page) { mutableListOf() }.add(entry.cachedItem)
        MENU_MAX_PAGE = maxOf(MENU_MAX_PAGE, entry.page)
    }

    private fun shouldDisplayOnPage(action: MenuAction, page: Int): Boolean = when (action) {
        MenuAction.NEXT_PAGE -> page < MENU_MAX_PAGE
        MenuAction.PREVIOUS_PAGE -> page > 1
        else -> true
    }

    private fun resolveTargetPage(action: MenuAction, page: Int): Int = when (action) {
        MenuAction.NEXT_PAGE -> page + 1
        MenuAction.PREVIOUS_PAGE -> page - 1
        else -> page
    }

    private fun parseMaterial(name: String?, fallback: Material, id: String, plugin: AliienResize): Material {
        return name?.takeIf { it.isNotBlank() }
            ?.let { Material.matchMaterial(it.uppercase(Locale.ROOT)) }
            ?: run {
                plugin.logger.warning("Invalid material '$name' for '$id'. Using ${fallback.name}.")
                fallback
            }
    }

    private fun parseItemFlags(flags: List<String>?): List<ItemFlag> {
        return flags?.mapNotNull { flag ->
            runCatching { ItemFlag.valueOf(flag.uppercase(Locale.ROOT)) }.getOrNull()
        } ?: emptyList()
    }

    private fun replacePlaceholders(
        list: List<String>,
        price: String = "",
        scale: String = "",
        permission: String = "",
        suffix: String = "",
        targetPage: String = ""
    ): List<String> {
        return list.map { line ->
            line.replace("%price%", price)
                .replace("%scale%", scale)
                .replace("%permission%", permission)
                .replace("%suffix%", suffix)
                .replace("%target_page%", targetPage)
        }
    }

    private fun buildItem(material: Material, name: String, lore: List<String>, modelData: Int, flags: Array<ItemFlag>, glow: Boolean): ItemStack {
        return ItemBuilder(material)
            .name(name)
            .stringLore(lore)
            .customModelData(modelData)
            .addFlags(*flags)
            .glow(glow)
            .build()
    }
}